src/analysis/asymptotics.lean

```diff
@@ -0,0 +1,438 @@
+/-
+Copyright (c) 2019 Jeremy Avigad. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Jeremy Avigad
+
+We introduce these relations:
+
+  `is_bigo f g l`    : "f is big O of g along l"
+  `is_littleo f g l` : "f is little o of g along l"
+
+Here `l` is any filter on the domain of `f` and `g`, which are assumed to be the same. The codomains
+of `f` and `g` do not need to be the same; all that is needed that there is a norm associated with
+these types, and it is the norm that is compared asymptotically.
+
+Often the ranges of `f` and `g` will be the real numbers, in which case the norm is the absolute
+value. In general, we have
+
+  `is_bigo f g l ↔ is_bigo (λ x, ∥f x∥) (λ x, ∥g x∥) l`,
+
+and similarly for `is_littleo`. But our setup allows us to use the notions e.g. with functions
+to the integers, rationals, complex numbers, or any normed vector space without mentioning the
+norm explicitly.
+
+If `f` and `g` are functions to a normed field like the reals or complex numbers and `g` is always
+nonzero, we have
+
+  `is_littleo f g l ↔ tendsto (λ x, f x / (g x)) (nhds 0) l`.
+
+In fact, the right-to-left direction holds without the hypothesis on `g`, and in the other direction
+it suffices to assume that `f` is zero wherever `g` is. (This generalization is useful in defining
+the Fréchet derivative.)
+-/
+import analysis.normed_space.basic
+open filter
+
+namespace asymptotics
+
+variables {α : Type*} {β : Type*} {γ : Type*} {δ : Type*}
+
+section
+variables [has_norm β] [has_norm γ] [has_norm δ]
+
+def is_bigo (f : α → β) (g : α → γ) (l : filter α) : Prop :=
+∃ c, { x | ∥ f x ∥ ≤ c * ∥ g x ∥ } ∈ l.sets
+
+def is_littleo (f : α → β) (g : α → γ) (l : filter α) : Prop :=
+∀ c, { x | c * ∥ f x ∥ ≤ ∥ g x ∥ } ∈ l.sets
+
+theorem is_bigo.comp {f : α → β} {g : α → γ} {l : filter α} (hfg : is_bigo f g l)
+    {δ : Type*} (k : δ → α) :
+  is_bigo (f ∘ k) (g ∘ k) (l.comap k) :=
+let ⟨c, hfgc⟩ := hfg in
+⟨c, mem_comap_sets.mpr ⟨_, hfgc, set.subset.refl _⟩⟩
+
+theorem is_littleo.comp {f : α → β} {g : α → γ} {l : filter α} (hfg : is_littleo f g l)
+    {δ : Type*} (k : δ → α) :
+  is_littleo (f ∘ k) (g ∘ k) (l.comap k) :=
+λ c, mem_comap_sets.mpr ⟨_, hfg c, set.subset.refl _⟩
+
+theorem is_bigo.mono {f : α → β} {g : α → γ} {l₁ l₂ : filter α} (h : l₁ ≤ l₂)
```

Comment:
Might `@[monotonicity]` be used here?

Reply:
I have never used it. If you are sure it is appropriate, I will add it. Otherwise, I will experiment by adding a local attribute later and see if it works.